pkg/dependenciesdistributor/dependencies_distributor.go

```diff
@@ -476,27 +483,56 @@ func (d *DependenciesDistributor) listAttachedBindings(bindingNamespace, binding
 	return res, nil
 }
 
-func (d *DependenciesDistributor) removeScheduleResultFromAttachedBindings(bindingNamespace, bindingName string, attachedBindings []*workv1alpha2.ResourceBinding) error {
+func (d *DependenciesDistributor) removeScheduleResultFromAttachedBindings(resourceBinding *workv1alpha2.ResourceBinding, attachedBindings []*workv1alpha2.ResourceBinding) error {
 	if len(attachedBindings) == 0 {
 		return nil
 	}
 
-	bindingLabelKey := generateBindingDependedLabelKey(bindingNamespace, bindingName)
-
 	var errs []error
 	for index, binding := range attachedBindings {
-		delete(attachedBindings[index].Labels, bindingLabelKey)
-		updatedSnapshot := deleteBindingFromSnapshot(bindingNamespace, bindingName, attachedBindings[index].Spec.RequiredBy)
+		delete(attachedBindings[index].Labels, bindingDependedByUIDLabelKey)
+		updatedSnapshot := deleteBindingFromSnapshot(resourceBinding.Namespace, resourceBinding.Name, attachedBindings[index].Spec.RequiredBy)
 		attachedBindings[index].Spec.RequiredBy = updatedSnapshot
 		if err := d.Client.Update(context.TODO(), attachedBindings[index]); err != nil {
 			klog.Errorf("Failed to update binding(%s/%s): %v", binding.Namespace, binding.Name, err)
 			errs = append(errs, err)
 		}
 	}
 
+	if len(errs) == 0 {
+		errs = append(errs, d.removeFinalizer(resourceBinding))
+	}
 	return utilerrors.NewAggregate(errs)
 }
 
+// addFinalizer adds finalizer to ResourceBinding
+func (d *DependenciesDistributor) addFinalizer(resourceBinding *workv1alpha2.ResourceBinding) error {
+	if controllerutil.ContainsFinalizer(resourceBinding, util.DependenciesDistributorFinalizer) {
+		return nil
+	}
+	controllerutil.AddFinalizer(resourceBinding, util.DependenciesDistributorFinalizer)
+	err := d.Client.Update(context.TODO(), resourceBinding)
```

Comment:
we almost don't have to consider conflicts because the finalizers are always maintained by the controller, so we'd better use `patch` instead of `update` to reduce the request body size.